### PR TITLE
[miio] compiler warnings & spotbugs cleanup

### PR DESCRIPTION
**Pull request by marcelrv**
no new functionality, just eliminating many warnings

Signed-off-by: Marcel Verpaalen <marcel@verpaalen.com>

### CODE REVIEW

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/basic/MiIoBasicChannel.java [hunk 1]
```diff
@@ -25,102 +28,111 @@
...
     }
 
     public void setChannelGroup(String channelGroup) {
         this.channelGroup = channelGroup;
     }
 
     public List<MiIoDeviceAction> getActions() {
-        return miIoDeviceActions;
+        final @Nullable List<MiIoDeviceAction> miIoDeviceActions = this.miIoDeviceActions;
+        return (miIoDeviceActions != null) ? miIoDeviceActions : new ArrayList<MiIoDeviceAction>();
```
Comment (cpmeister) [inline bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/basic/MiIoBasicChannel.java]: If these aren't supposed to be mutable you should return a Collections.emptyList() instead.

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/basic/MiIoDatabaseWatchService.java [hunk 2]
```diff
@@ -109,14 +112,11 @@ private void populateDatabase() {
...
             try {
-                @Nullable
                 JsonObject deviceMapping = Utils.convertFileToJSON(db);
-                if (deviceMapping != null) {
-                    Gson gson = new GsonBuilder().serializeNulls().create();
-                    MiIoBasicDevice devdb = gson.fromJson(deviceMapping, MiIoBasicDevice.class);
-                    for (String id : devdb.getDevice().getId()) {
-                        workingDatabaseList.put(id, db);
-                    }
+                Gson gson = new GsonBuilder().serializeNulls().create();
```
Comment (cpmeister) [inline bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/basic/MiIoDatabaseWatchService.java]: This Gson instance should be reused, but I guess you can do that in another PR.

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/discovery/MiIoDiscovery.java [hunk 3]
```diff
@@ -71,36 +74,44 @@ public MiIoDiscovery() throws IllegalArgumentException {
     @Override
     protected void startBackgroundDiscovery() {
         logger.debug("Start Xiaomi Mi IO background discovery");
+        final @Nullable ScheduledFuture<?> miIoDiscoveryJob = this.miIoDiscoveryJob;
         if (miIoDiscoveryJob == null || miIoDiscoveryJob.isCancelled()) {
-            miIoDiscoveryJob = scheduler.scheduleWithFixedDelay(() -> discover(), 0, SEARCH_INTERVAL, TimeUnit.SECONDS);
+            this.miIoDiscoveryJob = scheduler.scheduleWithFixedDelay(() -> discover(), 0, SEARCH_INTERVAL,
```
Comment (cpmeister) [inline bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/discovery/MiIoDiscovery.java]:
```suggestion
            this.miIoDiscoveryJob = scheduler.scheduleWithFixedDelay(this::discover, 0, SEARCH_INTERVAL,
```

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoAbstractHandler.java [hunk 4]
```diff
@@ -58,32 +58,34 @@
...
     private final Logger logger = LoggerFactory.getLogger(MiIoAbstractHandler.class);
     protected MiIoDatabaseWatchService miIoDatabaseWatchService;
 
-    @NonNullByDefault
-    public MiIoAbstractHandler(Thing thing) {
+    public MiIoAbstractHandler(Thing thing, MiIoDatabaseWatchService miIoDatabaseWatchService) {
         super(thing);
         parser = new JsonParser();
+        this.miIoDatabaseWatchService = miIoDatabaseWatchService;
+        this.configuration = getConfigAs(MiIoBindingConfiguration.class);
```
Comment (cpmeister) [inline bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoAbstractHandler.java]: I don't think you should grab the configuration until the handler is initialized, especially since configurations can change.

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoVacuumHandler.java [hunk 5]
```diff
@@ -139,7 +201,8 @@ private boolean updateVacuumStatus(JsonObject statusData) {
         int fanLevel = statusData.get("fan_power").getAsInt();
         updateState(CHANNEL_FAN_POWER, new DecimalType(fanLevel));
         updateState(CHANNEL_FAN_CONTROL, new DecimalType(FanModeType.getType(fanLevel).getId()));
-        updateState(CHANNEL_IN_CLEANING, new DecimalType(statusData.get("in_cleaning").getAsBigDecimal()));
+        inCleaning = statusData.get("in_cleaning").getAsInt();
+        updateState(CHANNEL_IN_CLEANING, new DecimalType(fanLevel));
```
Comment (cpmeister) [inline bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoVacuumHandler.java]: fanLevel?

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/transport/MiIoAsyncCommunication.java [hunk 6]
```diff
@@ -191,19 +195,22 @@ MiIoSendCommand sendMiIoSendCommand(MiIoSendCommand miIoSendCommand) {
     }
 
     public synchronized void startReceiver() {
+        MessageSenderThread senderThread = this.senderThread;
         if (senderThread == null) {
             senderThread = new MessageSenderThread();
         }
         if (!senderThread.isAlive()) {
             senderThread.start();
         }
```
Comment (cpmeister) [inline bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/transport/MiIoAsyncCommunication.java]: You can reuse threads like this?

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/basic/MiIoBasicChannel.java (hunk 1 above)
Comment (marcelrv) [reply bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/basic/MiIoBasicChannel.java]: thanks.. changed it.

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/basic/MiIoDatabaseWatchService.java (hunk 2 above)
Comment (marcelrv) [reply bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/basic/MiIoDatabaseWatchService.java]: Thanks, moved it so it is reused indeed.

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/discovery/MiIoDiscovery.java (hunk 3 above)
Comment (marcelrv) [reply bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/discovery/MiIoDiscovery.java]: thanks. updated it to your suggestion

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoAbstractHandler.java (hunk 4 above)
Comment (marcelrv) [reply bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoAbstractHandler.java]: ok, makes sense. removed it

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoVacuumHandler.java (hunk 5 above)
Comment (marcelrv) [reply bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoVacuumHandler.java]: :-) good catch... was doing bit trail & error testing.. seems I did not change this one back.

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/transport/MiIoAsyncCommunication.java (hunk 6 above)
Comment (marcelrv) [reply bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/transport/MiIoAsyncCommunication.java]: Not 100% sure. I changed it to be sure.

**Comment by marcelrv** [comment]
Thanks @cpmeister for your feedback. Made all suggested changes.

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/basic/MiIoBasicChannel.java [hunk 7]
```diff
@@ -25,102 +29,111 @@
...
     public void setChannel(String channel) {
         this.channel = channel;
     }
 
     public String getChannelType() {
-        return channelType == null || channelType.isEmpty() ? BINDING_ID + ":" + channel
-                : (channelType.startsWith("system") ? channelType : BINDING_ID + ":" + channelType);
+        final @Nullable String ct = channelType;
+        return ct == null || ct.isEmpty() ? BINDING_ID + ":" + channel
+                : (ct.startsWith("system") ? ct : BINDING_ID + ":" + ct);
```
Comment (cpmeister) [inline bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/basic/MiIoBasicChannel.java]: too many conditional operators and `":"` mixed with `:`. Can you make this more readable?

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoAbstractHandler.java [hunk 8]
```diff
@@ -335,9 +336,11 @@ protected boolean initializeData() {
 
     /**
      * Prepares the ExpiringCache for network data
+     *
+     * @return
      */
-    protected void initalizeNetworkCache() {
-        network = new ExpiringCache<String>(CACHE_EXPIRY_NETWORK, () -> {
+    protected ExpiringCache<String> initalizeNetworkCache() {
+        final ExpiringCache<String> network = new ExpiringCache<String>(CACHE_EXPIRY_NETWORK, () -> {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoAbstractHandler.java]: Why can't this created in the constructor? That way you can make the `network` field non-null and final.

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/transport/MiIoAsyncCommunication.java [hunk 9]
```diff
@@ -191,12 +195,11 @@ MiIoSendCommand sendMiIoSendCommand(MiIoSendCommand miIoSendCommand) {
...
 
     public synchronized void startReceiver() {
-        if (senderThread == null) {
+        MessageSenderThread senderThread = this.senderThread;
+        if (senderThread == null || !senderThread.isAlive()) {
             senderThread = new MessageSenderThread();
         }
-        if (!senderThread.isAlive()) {
-            senderThread.start();
-        }
```
Comment (cpmeister) [inline bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/transport/MiIoAsyncCommunication.java]: If you create a new thread isn't it still supposed to be started here? right now you are starting the thread in the constructor (and elsewhere maybe, I didn't check too thoroughly).

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java [hunk 10]
```diff
@@ -394,36 +413,39 @@ private void updatePropsFromJsonObject(MiIoSendCommand response) {
...
+        }
+        final String transformation = basicChannel.getTransfortmation();
+        if (transformation != null) {
+            JsonElement transformed = Conversions.execute(transformation, val);
+            logger.debug("Transformed with '{}': {} {} -> {} ", transformation, basicChannel.getFriendlyName(), val,
+                    transformed);
+            val = transformed;
+        }
+        try {
+            if (basicChannel.getType().equals("Number")) {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java]: looks like a good place to use a switch statement.

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java [hunk 11]
```diff
@@ -191,11 +194,17 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
+        final MiIoBindingConfiguration configuration = getConfigAs(MiIoBindingConfiguration.class);
         logger.debug("Periodic update for '{}' ({})", getThing().getUID().toString(), getThing().getThingTypeUID());
+        final MiIoAsyncCommunication miioCom = this.miioCom;
         try {
             if (!hasConnection() || skipUpdate()) {
                 return;
             }
+            if (miioCom == null) {
+                initializeData();
+                return;
```
Comment (cpmeister) [inline bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java]: Why return here?

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java [hunk 12]
```diff
@@ -394,36 +413,39 @@ private void updatePropsFromJsonObject(MiIoSendCommand response) {
...
+                        val.getAsString().toLowerCase().equals("on") || val.getAsString().toLowerCase().equals("true")
+                                ? OnOffType.ON
+                                : OnOffType.OFF);
+            }
+            if (basicChannel.getType().equals("Color")) {
+                Color rgb = new Color(val.getAsInt());
+                HSBType hsb = HSBType.fromRGB(rgb.getRed(), rgb.getGreen(), rgb.getBlue());
+                updateState(basicChannel.getChannel(), hsb);
+            }
+        } catch (Exception e) {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java]: Any specific exceptions you are trying to catch? If so please list them instead of catching a generic exception.

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoAbstractHandler.java (hunk 8 above)
Comment (marcelrv) [reply bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoAbstractHandler.java]: yes, that's a good way to solve it. Done.

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java (hunk 10 above)
Comment (marcelrv) [reply bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java]: yes, that makes the code cleaner... done

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java (hunk 11 above)
Comment (marcelrv) [reply bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java]: yes, indeed not needed. Fixed

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java (hunk 12 above)
Comment (marcelrv) [reply bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java]: There were various updating issues (e.g. conversion issues, colors not updating ell etc)
There should be no particular errors cuurently known to happen here

Comment (cpmeister) [reply bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoBasicHandler.java]: leave as is for now then

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/basic/MiIoBasicChannel.java (hunk 7 above)
Comment (marcelrv) [reply bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/basic/MiIoBasicChannel.java]: I improved this now. I think it is more clear

**Comment by marcelrv** [comment]
Thanks once more @cpmeister for your feedback. Made suggested changes.

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoVacuumHandler.java [hunk 13]
```diff
@@ -51,19 +52,63 @@
...
 public class MiIoVacuumHandler extends MiIoAbstractHandler {
     private final Logger logger = LoggerFactory.getLogger(MiIoVacuumHandler.class);
-
     private ExpiringCache<String> status;
     private ExpiringCache<String> consumables;
     private ExpiringCache<String> dnd;
     private ExpiringCache<String> history;
-    private String lastHistoryId;
+    private String lastHistoryId = "";
+    private int inCleaning;
```
Comment (cpmeister) [inline bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoVacuumHandler.java]: Why was this made a field?

Comment (marcelrv) [reply bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoVacuumHandler.java]: Indeed... Not really for this PR... For the next one...
I extracted these changes from the upcoming big pr to separate these many cleanup changes from the real functionality changes.

https://community.openhab.org/t/xiaomi-robot-vacuum-binding/31317/1321
